### PR TITLE
[Add] Allow multiple optimizers and add extra joint priors

### PR DESCRIPTION
**Pull request by wqyin**
1. Allow multiple optimizers for different parameters
2. Map losses only to update relevant parameters
3. Add grad clipping to avoid `nan` loss in optimization
4. Add joint angle priors for A-pose
5. Add joint angle priors for foot, for insufficient foot keypoints
6. Fix joint angle limits for shoulder

### CODE REVIEW

File: xrmocap/model/loss/prior_loss.py [hunk 1]
```diff
@@ -179,6 +193,47 @@ def forward(self,
                 torch.tensor([1., -1., -1, -1.], device=body_pose.device)) -
                                 1)**2
 
+        if self.lock_foot:
```
Comment (LazyBusyYang) [inline xrmocap/model/loss/prior_loss.py:196]: What about merge the three `if`s into a for loop?
```python
parts_joint_prior_losses = []
pred_poses = []
limits = []
weights = []
if self.use_full_body:
    pred_poses.append(aa_to_sja(body_pose_reshape, self.R_t, self.R_t_inv))
    limits.append(self.sja_limits)
    weights.append(1.0)
if self.lock_foot:
    pred_poses.append(body_pose_sja[:, [6, 7, 9, 10]])
    limits.append(self.sja_lock_foot)
    weights.append(self.lock_foot_loss_weight)
...
for idx, weight in enumerate(weights):
    lower_limits = limits[idx][:, :, 0]
    upper_limits = limits[idx][:, :, 1]
    ...
    parts_joint_prior_losses.append(weight * loss)


```

File: xrmocap/model/registrant/smplify.py [hunk 2]
```diff
@@ -352,44 +356,104 @@ def __optimize_stage__(self,
...
+            # update with the default optimizer or ignore otherwise
+            # | {key}_opt_config |  fit_{key}  |       optimizer     |
+            # | -----------------| ------------| --------------------|
+            # |       True       |     True    |   {key}_optimizer   |
+            # |       False      |     True    |   default_optimizer |
+            # |       True       |     False   |        ignore       |
+            # |       False      |     False   |        ignore       |
+            self.individual_optimizer = True
+            _optim_param = optim_param.copy()
+            for key in _optim_param.copy().keys():
```
Comment (LazyBusyYang) [inline xrmocap/model/registrant/smplify.py:381]: Too many `copy()`. If the copy here is to prevent pop Error, you can try this.
```suggestion
            for key in list(_optim_param.keys()):
```

File: xrmocap/model/registrant/smplify.py [hunk 3]
```diff
@@ -352,44 +356,104 @@ def __optimize_stage__(self,
...
+                parameters = OptimizableParameters()
+                fit_flag = kwargs.pop(f'fit_{key}', True)
```
Comment (LazyBusyYang) [inline xrmocap/model/registrant/smplify.py:383]: if `fit_betas` is absent in kwargs, fit_flag will be True by default. I guess a default False is better than True?

File: xrmocap/model/registrant/smplify.py [hunk 4]
```diff
@@ -352,44 +356,104 @@ def __optimize_stage__(self,
...
+                else:
+                    self.logger.info(f'No optimizer defined for {key}, '
+                                     'get the default optimizer')
+
+            if len(_optim_param) > 0:
+                parameters = OptimizableParameters()
+                if 'default_optimizer' not in self.optimizer:
+                    self.logger.error(
+                        'Individual optimizer mode is selected but '
+                        'some optimizaters are not defined. '
```
Comment (LazyBusyYang) [inline xrmocap/model/registrant/smplify.py:402]: optimizaters

File: xrmocap/model/registrant/smplify.py [hunk 5]
```diff
@@ -352,44 +356,104 @@ def __optimize_stage__(self,
...
+                        'Please set the default_optimzier or set optimizer '
+                        f'for {_optim_param.keys()}.')
```
Comment (LazyBusyYang) [inline xrmocap/model/registrant/smplify.py:404]: Always remember raise an error after `self.logger.error()`.
`raise KeyError`

File: xrmocap/model/registrant/smplify.py [hunk 6]
```diff
@@ -352,44 +356,104 @@ def __optimize_stage__(self,
...
+                else:
+                    for key in _optim_param.copy().keys():
```
Comment (LazyBusyYang) [inline xrmocap/model/registrant/smplify.py:406]: Same as https://github.com/openxrlab/xrmocap/pull/56/files#r1051822502

File: xrmocap/model/registrant/smplify.py [hunk 7]
```diff
@@ -352,44 +356,104 @@ def __optimize_stage__(self,
...
+                        fit_flag = kwargs.pop(f'fit_{key}', True)
+                        value = _optim_param.pop(key)
+                        if fit_flag:
+                            parameters.add_param(
+                                key=key, param=value, fit_param=fit_flag)
+                    optimizers['default_optimizer'] = build_optimizer(
+                        parameters, self.optimizer['default_optimizer'])
 
         pre_loss = None
```
Comment (LazyBusyYang) [inline xrmocap/model/registrant/smplify.py:415]: Or `previous_loss`?

File: xrmocap/model/registrant/smplify.py [hunk 8]
```diff
@@ -637,6 +705,28 @@ def __compute_loss__(self,
...
+
+        Args:
+            losses (dict): Original loss, use handler_key as keys.
+
+        Returns:
+            dict: Processed loss, use parameter names as keys.
+                Original keys included.
+        """
+
+        for loss_key in losses.copy().keys():
```
Comment (LazyBusyYang) [inline xrmocap/model/registrant/smplify.py:719]:
```suggestion
        for loss_key in losses.keys():
```

File: xrmocap/model/registrant/smplify.py [hunk 9]
```diff
@@ -637,6 +705,28 @@ def __compute_loss__(self,
 
         return losses
 
+    def __post_process_loss__(self, losses: dict, **kwargs) -> dict:
```
Comment (LazyBusyYang) [inline xrmocap/model/registrant/smplify.py:708]: According to [PEP8](https://peps.python.org/pep-0008/#descriptive-naming-styles), starting with a single under line is better. The style I used before is unprofessional, and we can correct it from now on.
```suggestion
    def _post_process_loss(self, losses: dict, **kwargs) -> dict:
```

File: xrmocap/model/registrant/smplify.py (hunk 8 above)
Comment (wqyin) [reply xrmocap/model/registrant/smplify.py]: Fixed in [cab8731](https://github.com/openxrlab/xrmocap/pull/56/commits/cab87319d90d6f7f6e5f611602f0b4c4d7d6f938).

File: xrmocap/model/registrant/smplify.py (hunk 7 above)
Comment (wqyin) [reply xrmocap/model/registrant/smplify.py]: Fixed in [cab8731](https://github.com/openxrlab/xrmocap/pull/56/commits/cab87319d90d6f7f6e5f611602f0b4c4d7d6f938).

File: xrmocap/model/registrant/smplify.py (hunk 6 above)
Comment (wqyin) [reply xrmocap/model/registrant/smplify.py]: Fixed in [cab8731](https://github.com/openxrlab/xrmocap/pull/56/commits/cab87319d90d6f7f6e5f611602f0b4c4d7d6f938).

File: xrmocap/model/loss/prior_loss.py (hunk 1 above)
Comment (wqyin) [reply xrmocap/model/loss/prior_loss.py:189]: Refactored in [cab8731](https://github.com/openxrlab/xrmocap/pull/56/commits/cab87319d90d6f7f6e5f611602f0b4c4d7d6f938).

File: xrmocap/model/registrant/smplify.py (hunk 2 above)
Comment (wqyin) [reply xrmocap/model/registrant/smplify.py]: Fixed in [cab8731](https://github.com/openxrlab/xrmocap/pull/56/commits/cab87319d90d6f7f6e5f611602f0b4c4d7d6f938).

File: xrmocap/model/registrant/smplify.py (hunk 3 above)
Comment (wqyin) [reply xrmocap/model/registrant/smplify.py]: Fixed in [cab8731](https://github.com/openxrlab/xrmocap/pull/56/commits/cab87319d90d6f7f6e5f611602f0b4c4d7d6f938).

File: xrmocap/model/registrant/smplify.py (hunk 4 above)
Comment (wqyin) [reply xrmocap/model/registrant/smplify.py]: Fixed in [cab8731](https://github.com/openxrlab/xrmocap/pull/56/commits/cab87319d90d6f7f6e5f611602f0b4c4d7d6f938).

File: xrmocap/model/registrant/smplify.py (hunk 5 above)
Comment (wqyin) [reply xrmocap/model/registrant/smplify.py:404]: Fixed in [cab8731](https://github.com/openxrlab/xrmocap/pull/56/commits/cab87319d90d6f7f6e5f611602f0b4c4d7d6f938).

**Review by LazyBusyYang — APPROVED** [review]
